internal/venice-common/src/main/java/com/linkedin/venice/pubsub/manager/TopicMetadataFetcher.java

```diff
@@ -345,37 +307,91 @@ List<PubSubTopicPartitionInfo> getTopicPartitionInfo(PubSubTopic topic) {
-   * @throws PubSubTopicDoesNotExistException If the topic does not exist.
-   * @throws IllegalArgumentException If the partition number is negative.
-   * @throws VeniceException If the offset returned by the consumer is null.
-   *                         This could indicate a bug in the PubSubConsumerAdapter implementation.
-   * @throws PubSubOpTimeoutException If the consumer times out. This could indicate that the topic does not exist
-   *                         or the partition does not exist.
-   */
-  long getLatestOffset(PubSubTopicPartition pubSubTopicPartition) {
-    PubSubConsumerAdapter pubSubConsumerAdapter = acquireConsumer();
+  // Acquire/release wrapper
+  private <T> T withConsumer(Function<PubSubConsumerAdapter, T> fn) {
+    PubSubConsumerAdapter consumer = acquireConsumer();
     try {
-      long startTime = System.nanoTime();
-      Map<PubSubTopicPartition, Long> offsetMap = pubSubConsumerAdapter
-          .endOffsets(Collections.singleton(pubSubTopicPartition), getPubsubOffsetApiTimeoutDurationDefaultValue());
-      stats.recordLatency(GET_PARTITION_LATEST_OFFSETS, startTime);
-      Long offset = offsetMap.get(pubSubTopicPartition);
-      if (offset == null) {
-        LOGGER.error("Received null offset for topic-partition: {}", pubSubTopicPartition);
-        // This should never happen; if it does, it's a bug in the PubSubConsumerAdapter implementation
-        throw new VeniceException("Got null as latest offset for: " + pubSubTopicPartition);
-      }
-      return offset;
+      return fn.apply(consumer);
     } finally {
-      releaseConsumer(pubSubConsumerAdapter);
+      releaseConsumer(consumer);
     }
   }
 
+  // Get all partitions for a topic, record PARTITIONS_FOR once
+  private Collection<PubSubTopicPartition> listPartitionsForTopic(PubSubConsumerAdapter consumer, PubSubTopic topic) {
+
+    long t0 = System.nanoTime();
+    List<PubSubTopicPartitionInfo> infos = consumer.partitionsFor(topic);
+    stats.recordLatency(PARTITIONS_FOR, t0);
+
+    if (infos == null || infos.isEmpty()) {
+      LOGGER.warn("Topic: {} may not exist or has no partitions. Returning empty map.", topic);
+      return Collections.emptyList();
+    }
+    Collection<PubSubTopicPartition> parts = new HashSet<>(infos.size());
+    for (PubSubTopicPartitionInfo info: infos) {
+      parts.add(info.getTopicPartition());
+    }
+    return parts;
+  }
+
+  Map<PubSubTopicPartition, PubSubPosition> getEndPositionsForTopic(PubSubTopic topic) {
+    return withConsumer(consumer -> {
+      Collection<PubSubTopicPartition> parts = listPartitionsForTopic(consumer, topic);
+      if (parts.isEmpty()) {
+        return Collections.emptyMap();
+      }
+      long t0 = System.nanoTime();
+      Map<PubSubTopicPartition, PubSubPosition> res =
+          consumer.endPositions(parts, getPubsubOffsetApiTimeoutDurationDefaultValue());
+      stats.recordLatency(GET_TOPIC_LATEST_OFFSETS, t0);
```

Comment:
it should be END_POSITIONS
or t0 should be initialized at beginning of the call